### PR TITLE
Add NFS support for PVC metrics in controller

### PR DESCRIPTION
**Pull request by rensyct**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| dell/csm#94 |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- Ensured that controller logs provides health state of the filesystem
- Ensured that node logs provides health state of the filesystem along with volume stats

### CODE REVIEW

File: service/controller.go [hunk 1]
```diff
@@ -1806,14 +1806,78 @@ func (s *service) ControllerGetVolume(ctx context.Context,
...
+		fileAPI := gounity.NewFilesystem(unity)
+		isSnapshot := false
+		filesystem, err := fileAPI.FindFilesystemById(ctx, volID)
+
+		if err != nil {
+			var snapResp *types.Snapshot
+			snapshotAPI := gounity.NewSnapshot(unity)
+			snapResp, err = snapshotAPI.FindSnapshotById(ctx, volID)
+			if err != nil {
+				if err == gounity.FilesystemNotFoundError {
```
Comment (ashish2207) [inline service/controller.go:1824]: should check for snapshotnotfound error in case of findsnapshot error

**Review by ashish2207 — CHANGES_REQUESTED** [review]

File: service/controller.go [hunk 2]
```diff
@@ -1806,14 +1806,82 @@ func (s *service) ControllerGetVolume(ctx context.Context,
...
+			}
+			var snapResp *types.Snapshot
+			snapshotAPI := gounity.NewSnapshot(unity)
+			snapResp, err = snapshotAPI.FindSnapshotById(ctx, volID)
+			if err != nil {
+				if err == gounity.SnapshotNotFoundError {
+					abnormal = true
+					message = "filesystem not found"
+				}
+				return nil, status.Error(codes.NotFound, utils.GetMessageWithRunID(rid, "Find filesystem: %s failed with error: %v", volID, err))
```
Comment (ashish2207) [inline service/controller.go:1832]: if abnormal is set to true, this return should not happen

File: service/controller.go [hunk 3]
```diff
@@ -1806,14 +1806,82 @@ func (s *service) ControllerGetVolume(ctx context.Context,
...
+			}
+			isSnapshot = true
+
+			filesystem, err = s.getFilesystemByResourceID(ctx, snapResp.SnapshotContent.StorageResource.Id, arrayID)
```
Comment (ashish2207) [inline service/controller.go:1836]: should be skipped if abnormal is true

File: service/controller.go [hunk 4]
```diff
@@ -1806,14 +1806,82 @@ func (s *service) ControllerGetVolume(ctx context.Context,
...
+			if err != nil {
+				return nil, err
+			}
+		}
+		nfsShareID := ""
+		for _, nfsShare := range filesystem.FileContent.NFSShare {
```
Comment (ashish2207) [inline service/controller.go:1842]: should be skipped if abnormal true

File: service/controller.go [hunk 5]
```diff
@@ -1806,14 +1806,82 @@ func (s *service) ControllerGetVolume(ctx context.Context,
...
+			for _, host := range readOnlyRootHosts {
+				hosts = append(hosts, host.ID)
+			}
+			for _, host := range readWriteRootHosts {
+				hosts = append(hosts, host.ID)
+			}
+		}
+
+		// check if volume is in ready state
+		if filesystem.FileContent.Health.Value != 5 {
```
Comment (ashish2207) [inline service/controller.go:1878]: should be skipped if abnormal true

**Review by ashish2207 — CHANGES_REQUESTED** [review]

File: service/controller.go [hunk 6]
```diff
@@ -1806,14 +1806,82 @@ func (s *service) ControllerGetVolume(ctx context.Context,
...
-		message = "Volume is in ok state"
+		message = "volume is in ok state"
 	} else {
+		fileAPI := gounity.NewFilesystem(unity)
+		isSnapshot := false
+		filesystem, err := fileAPI.FindFilesystemById(ctx, volID)
+
+		if err != nil {
+			if err == gounity.FilesystemNotFoundError {
+				abnormal = true
```
Comment (ashish2207) [inline service/controller.go:1821]: if pvc not found as filesystem then we will set abnormal as true, but if it is found as snapshot, at line 1826, then we should not say abnormal true

**Review by ashish2207 — CHANGES_REQUESTED** [review]

File: service/controller.go (hunk 1 above)
Comment (rensyct) [reply service/controller.go]: done

**Review by gallacher — CHANGES_REQUESTED** [review]
I do not see any unit tests new/updated to test the code that has been added here.  Please add those to this PR.

File: service/controller.go (hunk 2 above)
Comment (rensyct) [reply service/controller.go]: done

File: service/controller.go (hunk 3 above)
Comment (rensyct) [reply service/controller.go]: done

File: service/controller.go (hunk 4 above)
Comment (rensyct) [reply service/controller.go]: done

File: service/controller.go (hunk 5 above)
Comment (rensyct) [reply service/controller.go]: done

File: service/controller.go (hunk 6 above)
Comment (rensyct) [reply service/controller.go]: done

**Review by francis-nijay — DISMISSED** [review]

**Review by ashish2207 — APPROVED** [review]

**Review by prasar6 — DISMISSED** [review]

**Review by francis-nijay — DISMISSED** [review]

**Review by gallacher — APPROVED** [review]
Thank you for adding unit tests

**Review by prasar6 — APPROVED** [review]

**Pull request merged**